shared/src/business/entities/JoiValidationEntity.ts

```diff
@@ -0,0 +1,238 @@
+        }
+      });
+    } else if (typeof value === 'object' && value !== null) {
+      obj[key] = toRawObject(value);
+    } else {
+      obj[key] = value;
+    }
+  }
+  if (entity.isValidated) {
+    setIsValidated(obj);
+  }
+  return obj;
+}
+
+/**
+ * returns all of the validation errors after being converted to their formatted output
+ *
+ * @returns {object} the formatted errors
+ */
+function getFormattedValidationErrorsHelper(entity) {
+  const errors = entity.getValidationErrors();
+  if (!errors) return null;
+  for (let key of Object.keys(errors)) {
+    const errorMap = entity.getErrorToMessageMap()[key];
+    if (Array.isArray(errorMap)) {
+      for (let errorObject of errorMap) {
+        if (
+          typeof errorObject === 'object' &&
+          errors[key].includes(errorObject.contains)
+        ) {
+          errors[key] = errorObject.message;
+          break;
+        } else if (typeof errorObject !== 'object') {
+          errors[key] = errorObject;
+          break;
+        }
+      }
+    } else if (errorMap) {
+      errors[key] = errorMap;
+    }
+  }
+  return errors;
+}
+
+/**
+ * returns all of the validation errors after being converted to their formatted output
+ *
+ * @returns {object} the formatted errors
+ */
+function getFormattedValidationErrors(entity) {
+  const keys = Object.keys(entity);
+  const obj = {};
+  let errors = null;
+  if (entity.getFormattedValidationErrors) {
+    errors = getFormattedValidationErrorsHelper(entity);
+  }
+  if (errors) {
+    for (const key of Object.keys(errors)) {
+      if (
+        // remove unhelpful error messages from contact validations
+        typeof errors[key] == 'string' &&
+        errors[key].endsWith('does not match any of the allowed types')
+      ) {
+        delete errors[key];
+      }
+    }
+    Object.assign(obj, errors);
+  }
+  for (let key of keys) {
+    const value = entity[key];
+    if (errors && errors[key]) {
+      continue;
+    } else if (Array.isArray(value)) {
+      obj[key] = value
+        .map((v, index) => {
+          const e = getFormattedValidationErrors(v);
+          return e ? { ...e, index } : null;
+        })
+        .filter(v => v);
+      if (obj[key].length === 0) {
+        delete obj[key];
+      }
+    } else if (
+      typeof value === 'object' &&
+      value &&
+      value.getFormattedValidationErrors
+    ) {
+      obj[key] = getFormattedValidationErrors(value);
+      if (!obj[key]) delete obj[key];
+    }
+  }
+  return Object.keys(obj).length === 0 ? null : obj;
+}
+
+export abstract class JoiValidationEntity {
+  public entityName: string;
+  private schema: any;
+
+  constructor(entityName: string) {
+    this.entityName = entityName;
+    const rules = this.getValidationRules();
+
+    Object.defineProperty(this, 'schema', {
+      enumerable: false,
+      value: joi.object().keys(rules),
+      writable: true,
+    });
+  }
+
+  abstract getValidationRules(): any;
```

Comment:
Is there any reason we need to have a function which fetches the validation rules? Instead could we just set the schema property directly? 

So schema would be an abstract property that all classes implementing JoiValidationEntity would set on their own.

Reply:
when you put the schema directly on the object, it tries to get stored to dynamodb when we persist.  We could potentially have the schema be returned from a getter function, but I wanted to keep it off of the object as a public field.  You can't use abstract private in typescript.